### PR TITLE
Patch `send_lines` exploit

### PR DESCRIPTION
**Pull request by lokka30**
# About

> Note: This should be merged alongside #721 to fully patch the two exploits mentioned in #624.

Piqueserver does not prevent users from spamming commands which utilize the `send_lines` function. For commands that send a relatively larger amount of lines in the `send_lines` function, such as `/commands`, this exacerbates the issue.

This pull request does not deal with command spam (which has been left to #624) – this pull request *does* deal with multiple `send_lines` calls of the same underlying reason.

# Explanation of implementation

To address this exploit, I have added a set "`current_send_lines_types`" to `FeatureConnection`, which stores the list of types that a player is currently being sent.

Types are simply 'reasons' that `send_lines` has been called, which serve as identifiers, so that `send_lines` can be called only once at a time for a particular message.

For example, running `/commands` uses the type `commands` for the `send_lines` function.

If `send_lines` is called but the user is already receiving the lines of the given `type`, then an `info`-level message is logged to the console to inform the server that the function call has been silently discarded.

# To-do

- [x] Initial commit to fix exploit
- [x] Maintainer's satisfaction with the direction of this PR
- [x] Document the `type` parameter on `send_lines`
- [x] Mark pull request as 'ready to merge'

### CODE REVIEW

File: piqueserver/player.py [hunk 1]
```diff
@@ -333,12 +335,25 @@ def ban(self, reason=None, duration=None):
                 self.protocol.add_ban(self.address[0], reason, duration,
                                       self.name)
 
-    def send_lines(self, lines: List[str]) -> None:
+    def send_lines(self, lines: List[str], type: str = 'unknown') -> None:
```
Comment (NotAFile) [inline piqueserver/player.py:338]: I would rename type to key, as type is ambiguous and collides with the built-in type function

File: piqueserver/player.py [hunk 2]
```diff
@@ -43,6 +44,7 @@ def __init__(self, *args, **kwargs):
         self.user_types = None
         self.rights = None
         self.can_complete_line_build = True
+        self.current_send_lines_types = {}
```
Comment (NotAFile) [inline piqueserver/player.py:47]: `{}` creates a dict, not a set, so this will not run

**Review by NotAFile — CHANGES_REQUESTED** [review]
Overall I am happy with this approach.

There's a few other things that could be done, like only using a single task, cancelling the previous send_lines if a new one appears, but this should be good enough for now.

**Comment by lokka30** [comment]
Thanks for checking it out nota. I've attempted to resolve your suggestions, along the way I've also added a comment for the function and made it skip any detection if the default/unspecified key is used.

I have a thought that it might be better to add a function with the old syntax back (only the `lines` parameter), make that deprecated, and remove the default key on the new function (so `key: str` instead of `key: str='unknown'`). The 'old' function will then just call the new one using the default key.

**Pull request merged**